### PR TITLE
Main Menu, Players Setup Ships, Winner, End Game

### PR DESCRIPTION
**Pull request by Arique1104**
Alrighty!  So I need to see if this look is beneficial and there is tons of refactoring that needs doing, but I would like to merge this so that we can get your awesome methods for firing, so that we can integrate.

### CODE REVIEW

**Comment by priyapower** [comment]
Alrighty! So I need to see if this look is beneficial and there is tons of refactoring that needs doing, but I would like to merge this so that we can get your awesome methods for firing, so that we can integrate. **We can absolutely work on this merge to dev2 so we can get the methods interacting with each other."

File: lib/game.rb [hunk 1]
```diff
@@ -34,14 +34,66 @@ def message_player_display
...
   # ============================================
 
-  def fire(declarer, coordinate)
-    if declarer == @user
-      # nested if for user error for firing on coordinate that has been fired on
-      @auto.board.cells[coordinate].fire_upon
-      # require "pry"; binding.pry
-    elsif declarer == @auto
+  def start
+    main_menu
```
Comment (priyapower) [inline lib/game.rb:38]: Love that the start method just immediately has so much accessibility for the user.

File: lib/game.rb [hunk 2]
```diff
@@ -34,14 +34,66 @@ def message_player_display
...
+
+    if response == "q"
+      print "Good Bye!"
+      leave_game
+    elsif response == "p"
+      print "Let's play!"
+    else response != "p" || "q"
+      print "Invalid entry, please enter p to play or q to quit. > "
+      response2 = gets.chomp.downcase[0]
+      if response2 == "p"
```
Comment (priyapower) [inline lib/game.rb:60]: If the loops from fire coordinates are working for user input, we can use something similar during the refactor.

File: lib/game.rb [hunk 3]
```diff
@@ -34,14 +34,66 @@ def message_player_display
...
+    # results of winner
+    #fire
     end
+    end_game
+  end
+
+  def main_menu
+    print "Welcome to BATTLESHIP Enter p to play. Enter q to quit. > "
+
+    response = gets.chomp.downcase[0]
```
Comment (priyapower) [inline lib/game.rb:50]: Love the guard here to cut off users extra character input

File: lib/game.rb [hunk 4]
```diff
@@ -34,14 +34,66 @@ def message_player_display
...
+        leave_game
+      else response2 != "p" || "q"
+        print "Good- bye!"
+        leave_game
+      end
+    end
+
+  end
+
+  def leave_game
```
Comment (priyapower) [inline lib/game.rb:73]: I like that this method holds the exit. This allows for readability about what exit is actually doing.

File: lib/game.rb [hunk 5]
```diff
@@ -34,14 +34,66 @@ def message_player_display
...
+    exit
+  end
+
+  def players_setup_ships
```
Comment (priyapower) [inline lib/game.rb:77]: Love this! Can't wait to further edge case test this (we did quite a lot in the player class already) once the game is rendering correctly :)

File: lib/game.rb [hunk 6]
```diff
@@ -34,14 +34,66 @@ def message_player_display
...
+    @auto.ship_setup
+    @user.ship_setup
+  end
+
+
+
+  def winner
```
Comment (priyapower) [inline lib/game.rb:84]: Fantastic logic on the winner method. Love that this is adaptable for dynamic use later is players change number of ships or board size.

File: lib/game.rb [hunk 7]
```diff
@@ -34,14 +34,66 @@ def message_player_display
...
+    if @auto.ships.sum {|ship| ship.health } == 0
+      @user
+    elsif @user.ships.sum { |ship| ship.health } == 0
+      @auto
+    else
+      nil
+    end
+  end
 
+  def end_game
```
Comment (priyapower) [inline lib/game.rb:94]: Woo! There is the connection back to main menu! Love it!

File: lib/player.rb [hunk 8]
```diff
@@ -82,6 +82,12 @@ def all_ships_placed?
     ship_count == ships_on_board
   end
 
+  def cell_fired_upon?(coord)
```
Comment (priyapower) [inline lib/player.rb:85]: Love that you wrote this in. I think I may be able to use this in the refactor for the fire_coordinate (which is for user input or technically auto input of firing coordinate)

File: lib/ship.rb [hunk 9]
```diff
@@ -10,15 +10,11 @@ def initialize(name, length)
...
 
   def hit
-      @health -= 1
+    @health -= 1
   end
 
   def sunk?
-    if @health == 0
-      true
-    else
```
Comment (priyapower) [inline lib/ship.rb:19]: Is this else to prevent a nil?

File: test/player_test.rb [hunk 10]
```diff
@@ -92,4 +92,10 @@ def test_it_can_auto_generate_coordinates
 
   end
 
+  def test_if_the_cell_was_fired_upon
```
Comment (priyapower) [inline test/player_test.rb:95]: thanks for writing a matching test for your updated method!

File: test/game_test.rb [hunk 11]
```diff
@@ -75,4 +75,87 @@ def test_DEBUG_board_renders_shot_behavior_for_testing
 
   end
 
+  def test_it_has_a_main_menu
```
Comment (priyapower) [inline test/game_test.rb:78]: Confirming I see a test for start --> main menu method

File: test/game_test.rb [hunk 12]
```diff
@@ -75,4 +75,87 @@ def test_DEBUG_board_renders_shot_behavior_for_testing
...
+    skip
+    #does print to the screen as expected
+    expected = "Welcome to BATTLESHIP
+Enter p to play. Enter q to quit."
+
+    assert_equal expected, @game.main_menu
+
+  end
+
+  def test_it_has_a_set_up_ships
```
Comment (priyapower) [inline test/game_test.rb:88]: Confirming I see test for players_setup_ships

File: test/game_test.rb (hunk 11 above)
Comment (priyapower) [reply test/game_test.rb:78]: Do we need tests for start method by itself or leave_game or any helper methods written in?

File: test/game_test.rb [hunk 13]
```diff
@@ -75,4 +75,87 @@ def test_DEBUG_board_renders_shot_behavior_for_testing
...
+    auto.add_ship(auto_cruiser)
+    auto.add_ship(auto_submarine)
+    user.add_ship(user_cruiser)
+    user.add_ship(user_submarine)
+    game = Game.new(user, auto)
+
+    assert_equal nil, game.players_setup_ships
+  end
+
+  def test_it_can_determine_winner
```
Comment (priyapower) [inline test/game_test.rb:112]: Confirming I see test for method winner.

File: lib/game.rb [hunk 14]
```diff
@@ -34,14 +34,66 @@ def message_player_display
...
+
+  def players_setup_ships
+    @auto.ship_setup
+    @user.ship_setup
+  end
+
+
+
+  def winner
+    if @auto.ships.sum {|ship| ship.health } == 0
```
Comment (priyapower) [inline lib/game.rb:85]: Can these be helpers?

File: test/game_test.rb [hunk 15]
```diff
@@ -75,4 +75,87 @@ def test_DEBUG_board_renders_shot_behavior_for_testing
...
+
+    assert_equal 0, user.ships[1].health
+
+    # assert_equal true, game.all_ships_sunk?(user)
+    assert_equal auto, game.winner
+
+    #update with game.fire method to fully test functionality
+  end
+
+  def test_end_game_returns_to_main_menu
```
Comment (priyapower) [inline test/game_test.rb:156]: Confirming I see test for end game method.

File: lib/game.rb (hunk 6 above)
Comment (Arique1104) [reply lib/game.rb:84]: Thank you!  I thought since we had the @ships array, we should simply take the health integer sum and that can cut out the assumption that we would only have 5 total in health (1 cruiser and 1 submarine totaling 5).

File: lib/game.rb (hunk 7 above)
Comment (Arique1104) [reply lib/game.rb:94]: I'm out here! #Phew! #Glad it worked!

File: lib/player.rb (hunk 8 above)
Comment (Arique1104) [reply lib/player.rb:85]: I'm all for reusing code.  And I have some ideas about our render method now that you mention refactoring.  It might be nothing, but I was wondering why the cell AND the board BOTH have render methods that reference the same exact cells.  I put down a note that I think this would be a perfect moment for a module.

File: lib/ship.rb (hunk 9 above)
Comment (Arique1104) [reply lib/ship.rb:19]: Actually, on line 17, I put it back to @health == 0.  I probably did change that to prevent a nil, but honestly even with the change I don't believe I accounted for the nil.  @health is perfectly within reach of the ship scope.

File: test/game_test.rb (hunk 11 above)
Comment (Arique1104) [reply test/game_test.rb:78]: Yes, there are so many tests I need to do for this.  I'm not doing proper TDD.  But honestly, now that we've merged your awesome fire methods, I really feel like I can dig my teeth into this.  Also, most of the tests that would be written for these are focused on what is being printed to the terminal.  So per Tim's instruction, I am no longer going to test what gets printed to the terminal, but I do need to test functionality.

File: test/game_test.rb (hunk 12 above)
Comment (Arique1104) [reply test/game_test.rb:88]: Yes, I am writing a method so that within our Game.start we have the menu to set up the ship methods show up in our game class prompting both players to post their ships!

I'm not sure if it's the right call, but it sure makes it easier to have what is being printed onto the terminal so close to the functionality of the method.

File: lib/game.rb (hunk 6 above)
Comment (priyapower) [reply lib/game.rb:84]: <3

File: lib/game.rb (hunk 7 above)
Comment (priyapower) [reply lib/game.rb:94]: <3

File: lib/player.rb (hunk 8 above)
Comment (priyapower) [reply lib/player.rb:85]: Yessss! I had the same question. I'm glad we are both thinking through our work searching for repetitive moments!

File: lib/ship.rb (hunk 9 above)
Comment (priyapower) [reply lib/ship.rb:19]: Gotcha. Thanks for clarifying!

File: test/game_test.rb (hunk 11 above)
Comment (priyapower) [reply test/game_test.rb:78]: Love it. Yes, I agree. Most needs to be TDD to create, but I was also a little lost on how to TDD when working helper methods. I'm starting to see what is best-practice, but I would LOVE to chat and rework our git add commit flow for any attempt on Iteration 4.
